components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.java

```diff
@@ -0,0 +1,398 @@
+/*
+  Copyright (C) 2013-2018 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.client;
+
+import com.google.common.collect.ImmutableList;
+import com.hotels.styx.api.HttpRequest;
+import com.hotels.styx.api.HttpResponse;
+import com.hotels.styx.api.Id;
+import com.hotels.styx.api.extension.Origin;
+import com.hotels.styx.api.extension.RemoteHost;
+import com.hotels.styx.api.extension.loadbalancing.spi.LoadBalancer;
+import com.hotels.styx.api.extension.retrypolicy.spi.RetryPolicy;
+import com.hotels.styx.api.RequestCookie;
+import com.hotels.styx.api.exceptions.NoAvailableHostsException;
+import com.hotels.styx.api.HttpResponseStatus;
+import com.hotels.styx.api.MetricRegistry;
+import com.hotels.styx.api.metrics.codahale.CodaHaleMetricRegistry;
+import com.hotels.styx.api.extension.service.RewriteRule;
+import com.hotels.styx.api.extension.service.StickySessionConfig;
+import com.hotels.styx.client.retry.RetryNTimes;
+import com.hotels.styx.client.stickysession.StickySessionLoadBalancingStrategy;
+import com.hotels.styx.server.HttpInterceptorContext;
+import org.slf4j.Logger;
+import rx.Observable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Objects.toStringHelper;
+import static com.google.common.collect.Lists.newArrayList;
+import static com.hotels.styx.api.HttpHeaderNames.CONTENT_LENGTH;
+import static com.hotels.styx.api.HttpHeaderNames.TRANSFER_ENCODING;
+import static com.hotels.styx.api.StyxInternalObservables.toRxObservable;
+import static com.hotels.styx.api.extension.service.StickySessionConfig.stickySessionDisabled;
+import static com.hotels.styx.client.stickysession.StickySessionCookie.newStickySessionCookie;
+import static io.netty.handler.codec.http.HttpMethod.HEAD;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.nonNull;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.StreamSupport.stream;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * A configurable HTTP client that uses connection pooling, load balancing, etc.
+ */
+public final class StyxBackendServiceClient implements BackendServiceClient {
```

Comment:
It would be nice if this class could be named `BackendServiceClient`, though that would obvious conflict with the interface name.